src/python/dxpy/cli/dataset_utilities.py

```diff
@@ -0,0 +1,559 @@
+            print_to_stdout = True
+        elif os.path.exists(args.output):
+            if os.path.isdir(args.output):
+                out_directory = args.output
+                out_file_field = os.path.join(out_directory, resp['recordName'] + file_name_suffix)
+                files_to_check.append(out_file_field)
+            else:
+                file_already_exist.append(args.output)
+        elif os.path.exists(os.path.dirname(args.output)) or not os.path.dirname(args.output):
+            out_file_field = args.output
+        else:
+            err_exit(fill("Error: {path} could not be found".format(path=os.path.dirname(args.output))))
+
+    if files_to_check:
+        for file in files_to_check:
+            if os.path.exists(file):
+                file_already_exist.append(file)
+    
+    if file_already_exist:
+        err_exit(fill("Error: path already exists {path}".format(path=file_already_exist)))
+
+    rec_descriptor = DXDataset(dataset_id, project=dataset_project).get_descriptor()
+    if args.fields is not None:
+        fields_list = ''.join(args.fields).split(',')
+        error_list = []
+        for entry in fields_list:
+            entity_field = entry.split('.')
+            if len(entity_field) < 2 :
+                error_list.append(entry)
+            elif entity_field[0] not in rec_descriptor.model["entities"].keys() or \
+               entity_field[1] not in rec_descriptor.model["entities"][entity_field[0]]["fields"].keys():
+               error_list.append(entry)
+        
+        if error_list:
+            raise err_exit(fill('The following fields cannot be found: %r' % error_list))
+
+        payload = {"project_context":project, "fields":[{item:'$'.join(item.split('.'))} for item in fields_list]}
+        if "CohortBrowser" in resp['recordTypes']:
+            if resp.get('baseSql'):
+                payload['base_sql'] = resp.get('baseSql')
+            payload['filters'] = resp['filters']
+
+        if args.sql:
+            resource_val = resp['url'] + '/viz-query/3.0/' + resp['dataset'] + '/raw-query'
+            try:
+                resp_raw_query = dxpy.DXHTTPRequest(resource=resource_val, data=payload, prepend_srv=False)
+            except Exception as details:
+                raise ResolutionError(str(details))
+            sql_results = resp_raw_query['sql'] + ';'
+            if print_to_stdout:
+                print(sql_results)
+            else:
+                with open(out_file_field, 'w') as f:
+                    print(sql_results, file=f)
+        else:
+            resource_val = resp['url'] + '/data/3.0/' + resp['dataset'] + '/raw'
+            try:
+                resp_raw = dxpy.DXHTTPRequest(resource=resource_val, data=payload, prepend_srv=False)
+                if 'error' in resp_raw.keys():
+                    if resp_raw['error']['details'] and 'PermissionDenied: Access to the specified database is not allowed' in resp_raw['error']['details']:
```

Comment:
Would be great if we had specific error codes to target here. We currently only operate in english but should we ever introduce some localization to our error messages (or even just change the english wording) this check would fail.

Reply:
Agreed! We don't have a code for this error as of now. In this case, even if the error message gets changed somewhere, the else part will catch the error. But I totally agree that having error codes is much more cleaner.

Reply:
Agree with @jkotrs, we shouldn't rely on error outputs that are not a part of exception/error API. How about simply printing/throwing the error we got from the server?

Reply:
We can print the error message from server, but that will stop us from printing the error message defined in the spec to handle this error. @durae what do you suggest here?

Reply:
That error message is actually not correct in the condition, so let's do something similar to: 
```
if resp_raw['error']['type'] == 'InvalidInput':
    print("Insufficient permissions due to the project policy.")
    print(resp_raw['error']['message'])
```

Reply:
Updated!